src/gs-background-tile.c

```diff
@@ -0,0 +1,276 @@
+/* -*- Mode: C; tab-width: 8; indent-tabs-mode: t; c-basic-offset: 8 -*-
+ *
+ * Copyright (C) 2013 Matthias Clasen <mclasen@redhat.com>
+ * Copyright (C) 2017 Joaquim Rocha <jrocha@endlessm.com>
+ *
+ * Licensed under the GNU General Public License Version 2
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "config.h"
+
+#include <glib/gi18n.h>
+#include <gtk/gtk.h>
+
+#include "gs-background-tile.h"
+#include "gs-star-widget.h"
+#include "gs-common.h"
+
+struct _GsBackgroundTile
+{
+	GsAppTile	 parent_instance;
+
+	GsApp		*app;
+	GtkWidget	*name_label;
+	GtkWidget	*summary_label;
+	GtkWidget	*image;
+	GtkWidget	*image_box;
+	GtkWidget	*installed_icon;
+	GtkWidget	*price;
+	GtkWidget	*stack;
+};
+
+G_DEFINE_TYPE (GsBackgroundTile, gs_background_tile, GS_TYPE_APP_TILE)
+
+static GsApp *
+gs_background_tile_get_app (GsAppTile *tile)
+{
+	return GS_BACKGROUND_TILE (tile)->app;
+}
+
+static void
+gs_background_tile_setup_colors_bg (GsBackgroundTile *tile)
+{
+	GsApp *app = tile->app;
+	const guint num_colors = 4;
+	const GPtrArray *colors = gs_app_get_key_colors (app);
+	const gchar *gradients[num_colors];
+	g_autofree gchar *css = NULL;
+	gint degrees = 45;
+
+	if (colors->len == 0)
+		return;
+
+	/* apply one color gradient for each color from the cornors in the
+	 * following fashion:
+	 * -------  -------  -------  -------
+	 * -     -  - /   -  - / \ -  - / \ -
+	 * - \   -  - \   -  - \   -  - \ / -
+	 * -------  -------  -------  -------
+	 */
+	for (guint i = 0; i < num_colors; ++i) {
+		GdkRGBA *rgba = g_ptr_array_index (colors, i % colors->len);
+		guint red, green, blue;
+
+		/* we use the values in the printf below instead of using the
+		 * convenient gdk_rgba_to_string to avoid extra allocations as
+		 * this function can be used many times */
```

Comment:
I'm pretty sure that calling `gs_utils_widget_set_css()` will dominate any string allocations we can ever do here in terms of performance; I'd prefer to use the canonical `gdk_rgba_to_string()` unless you profile it to be an issue.